### PR TITLE
[SPARK-14410][SQL] Push functions existence check into catalog

### PR DESCRIPTION
**Pull request by andrewor14**
## What changes were proposed in this pull request?

This is a followup to #12117 and addresses some of the TODOs introduced there. In particular, the resolution of database is now pushed into session catalog, which knows about the current database. Further, the logic for checking whether a function exists is pushed into the external catalog.

No change in functionality is expected.
## How was this patch tested?

`SessionCatalogSuite`, `DDLSuite`

### CODE REVIEW

**Comment by andrewor14** [comment]
@yhuai

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala [hunk 1]
```diff
@@ -152,6 +152,8 @@ abstract class ExternalCatalog {
 
   def getFunction(db: String, funcName: String): CatalogFunction
 
+  def functionExists(db: String, funcName: String): Boolean
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: Great!

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 2]
```diff
@@ -559,7 +559,11 @@ private[hive] class HiveClientImpl(
   override def getFunctionOption(
       db: String,
       name: String): Option[CatalogFunction] = withHiveState {
-    Option(client.getFunction(db, name)).map(fromHiveFunction)
+    try {
+      Option(client.getFunction(db, name)).map(fromHiveFunction)
+    } catch {
+      case he: HiveException => None
```
Comment (yhuai) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: What will the Exception look like? No such function?

Comment (yhuai) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: Looks like the error that we want to return None is `org.apache.spark.sql.AnalysisException: org.apache.hadoop.hive.ql.metadata.HiveException: MetaException(message:NoSuchObjectException(message:Function default.example_max does not exist))`. Maybe try to matching the string (not sure if match HiveException.cause.cause is robust enough)?

**Comment by yhuai** [comment]
LGTM. Merging to master. We will take care the exception handling in getFunctionOption with the work of https://issues.apache.org/jira/browse/SPARK-14414.

**Pull request closed without merging**